Read only execution

gterzian opened the pull request:
Close #1977

damip commented:
@gterzian  updated spec: https://github.com/massalabs/massa/discussions/1970

gterzian commented:
@adrien-zinger @damip Ready for review, I'll do the API part in a separate PR.

adrien-zinger commented on massa-execution/src/vm.rs:
```diff
@@ -72,15 +75,15 @@ impl VM {
     ///
     /// # Parameters
     ///   * step: execution step to run
-    pub(crate) fn run_final_step(&mut self, step: &ExecutionStep) {
+    pub(crate) fn run_final_step(&mut self, step: ExecutionStep) {
         // check if that step was already executed as the earliest active step
-        let history_item = if let Some(cached) = self.pop_cached_step(step) {
+        let history_item = if let Some(cached) = self.pop_cached_step(&step) {
             // if so, pop it
             cached
         } else {
             // otherwise, clear step history an run it again explicitly
             self.step_history.clear();
-            self.run_step_internal(step)
+            self.run_step_internal(&step)
```
I guess that you can reset these lines and just borrow the step

gterzian replied:
I removed the borrow because it required borrowing in the run loop, which is weird because you actually want to take ownership of the values(for example in the new request to execute read-only stuff).

adrien-zinger commented on massa-execution/src/vm.rs:
```diff
@@ -174,6 +177,63 @@ impl VM {
+        address: Option<Address>,
+        result_sender: oneshot::Sender<ExecuteReadOnlyResponse>,
+    ) {
+        let saved_changes = {
+            let mut context = self.execution_context.lock().unwrap();
+
+            // Set the call stack, using the provided address, or a random one.
+            let address = address.unwrap_or_else(|| {
+                let private_key = generate_random_private_key();
+                let public_key = derive_public_key(&private_key);
+                Address::from_public_key(&public_key)
+            });
+            context.call_stack = vec![address].into();
+
+            // Set the max gas.
+            context.max_gas = max_gas;
+
+            // Set the simulated gas price.
+            context.gas_price = simulated_gas_price;
+
+            // Return a clone of the current state, to revert to it after execution.
+            context.ledger_step.caused_changes.clone()
+        };
```
It looks like something more generic, can we find a way to use the prepare_context or extract a function?

gterzian replied:
I don't think its' worth it in this PR. If you see a clear alternative feels free to open an issue/pr and propose the change.

damip replied:
we don't need to save/ pop changes at all actually :) just reset the context, run and that's it (just don't apply local_changes to the ledger at the end). The save/pop is there in run_step_internal because we need to keep some changes but not others, and we always apply at least some changes to the final ledger. Here it's not the case, so we can simplify :)

adrien-zinger commented on massa-models/src/execution.rs:
```diff
@@ -0,0 +1,22 @@
+use crate::Slot;
```
execution_*types*.rs?

gterzian replied:
I don't think so, since it's arleady in `models` it means it will be some sort of "types".

adrien-zinger commented on massa-execution/src/vm.rs:
```diff
@@ -174,6 +177,63 @@ impl VM {
+        // Send result back.
+        let execution_response = ExecuteReadOnlyResponse {
+            executed_at: slot,
+            // TODO: specify result.
+            result: run_result.map_or_else(
+                |_| ReadOnlyResult::Error("Failed to run in read-only mode".to_string()),
+                |_| ReadOnlyResult::Ok,
+            ),
```
We can return all the response or at least the `run_result.res`, what do you think?

gterzian replied:
I'm not sure, do we actually want to return anything other than the output events? @damip Looking at the latest version at https://github.com/massalabs/massa/discussions/1970#discussioncomment-1879779 it seems we don't.

adrien-zinger reviewed (commented):
It looks good enough for me. I just make some comments that can be fixed, ignored or turned into issues!

gterzian commented:
I will still integrate the spec in this PR(or do this in a follow-up if you want to merge this already today).

gterzian commented:
Actually let's merge this first(after approval from reviewers), since it makes it easier to link to the md file after it's been already checked in...

damip commented on massa-execution/src/vm.rs:
```diff
@@ -174,6 +177,63 @@ impl VM {
+            let mut context = self.execution_context.lock().unwrap();
+
+            // Set the call stack, using the provided address, or a random one.
+            let address = address.unwrap_or_else(|| {
```
the specification should be updated to reflect this random-address-by-default behavior

gterzian replied:
done

damip commented on massa-execution/src/vm.rs:
```diff
@@ -174,6 +177,63 @@ impl VM {
+        let run_result = assembly_simulator::run(&bytecode, max_gas, &*self.execution_interface);
+
+        // cancel the effects of execution only, pop back init_changes
+        self.execution_context
```
is not necessary to save/pop the context at all, since we  don't apply its results in the ledger.

Just reset it at the beginning, run the stuff, and forget it.

gterzian replied:
Ok thanks, done

damip commented on massa-execution/src/vm.rs:
```diff
@@ -174,6 +177,63 @@ impl VM {
+        result_sender: oneshot::Sender<ExecuteReadOnlyResponse>,
+    ) {
+        let saved_changes = {
+            let mut context = self.execution_context.lock().unwrap();
```
use self.clear_and_update_context(); (see the run_step_internal function) to initialize the context, otherwise it will simply be inherited from the last call. This way the rest of the context will also be reset to a default state here, to ensure that if we add fields to the context, there is no persistence between calls in case we forget to port them here

gterzian replied:
Ok thanks, done.

gterzian commented:
@damip @adrien-zinger Ok, thanks for the review and I've addressed the comments, please take another look and merge if ok. I'm adding the API in another branch.

damip reviewed (approved):
OK for me
